### PR TITLE
[go] Bump images, dependencies and versions to go 1.20

### PR DESCRIPTION
**Pull request by cpanato**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

- Bump images, dependencies and versions to go 1.20

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/release/issues/2815

#### Does this PR introduce a user-facing change?

```release-note
Kubernetes is now built with go 1.20
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by cpanato** [comment]
cc @dims @liggitt @kubernetes/release-managers

**Comment by cpanato** [comment]
@dims @liggitt any extra jobs that we need to run?

**Comment by dims** [comment]
/assign @liggitt 

cc @MadhavJivrajani

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -476,6 +476,7 @@ EOF
   local go_version
   IFS=" " read -ra go_version <<< "$(GOFLAGS='' go version)"
   local minimum_go_version
+  # TODO(cpanato): Need to switch this to 1.20 once we update images to newer go version
```
Comment (liggitt) [inline hack/lib/golang.sh:479]: also need to comment out the line in dependencies.yaml that is expecting this to be 1.20rc1

> `ERRO /home/prow/go/src/k8s.io/kubernetes/build/dependencies.yaml indicates that golang: 1.<major> should be at version 1.20rc1, but the following files didn't match: hack/lib/golang.sh`

**Comment by liggitt** [comment]
do we have canary CI jobs so we know that unit and integration tests pass on go1.20rc1 before merge?

**Comment by liggitt** [comment]
> do we have canary CI jobs so we know that unit and integration tests pass on go1.20rc1 before merge?

we do... now that the release-1.23 and release-1.24 go1.19 updates are merged, moving go-canary to go1.20rc1 would let us test unit / integration / dependencies / verify in presubmit before moving CI to go1.20rc1

**Comment by MadhavJivrajani** [comment]
Created https://github.com/kubernetes/test-infra/pull/28312

**Comment by MadhavJivrajani** [comment]
The `verify` canary job is failing due to go1.20.rc1 allowing [conversion from slice to array pointer](https://tip.golang.org/ref/spec#Conversions_from_slice_to_array_or_array_pointer). This is not yet supported in linters of golangci-lint. 
See https://github.com/golangci/golangci-lint/issues/3420
More specifically see https://github.com/dominikh/go-tools/issues/1343 and https://github.com/dominikh/go-tools/issues/1335 

They have fixed this in the master branch (https://github.com/dominikh/go-tools/commit/39b8ce77d7949d6d4383e1f9347ec50aefcb3573) and  hope to make it part of the next release (which will happen once go1.20 is released). See https://github.com/dominikh/go-tools/issues/1343#issuecomment-1356889094

**Comment by liggitt** [comment]
TestJitterWithNegativeMaxFactor is likely relying on math.Rand seeding to 1, which changes in go1.20:
> https://tip.golang.org/doc/go1.20#minor_library_changes
>
> The [math/rand](https://tip.golang.org/pkg/math/rand/) package now automatically seeds the global random number generator (used by top-level functions like Float64 and Int) with a random value, and the top-level [Seed](https://tip.golang.org/pkg/math/rand/#Seed) function has been deprecated. Programs that need a reproducible sequence of random numbers should prefer to allocate their own random source, using rand.New(rand.NewSource(seed)).

The other three failures look like something changed in the CA handling of VSphereConnection (or that the tests are looking for specific error strings the stdlib no longer returns)

**Comment by MadhavJivrajani** [comment]
> TestJitterWithNegativeMaxFactor is likely relying on math.Rand seeding to 1, which changes in go1.20:

+1, I agree!

The CA verification failures seem to be due to introduction of the `[CertificateVerificationError](https://tip.golang.org/pkg/crypto/tls/#CertificateVerificationError)` type in go1.20 and how errors are handled in `govmomi`. The text that is checked is still present in the package but it hasn't been updated since 2019. 

I believe this commit helps solve it: https://github.com/vmware/govmomi/commit/05b0b08c4997144accd4e4b94ee1ec1883c67623

I updated the version of this dep to the latest release (`v0.30.0`( and that seem to solve the issue. Will send a PR to update the version!

**Comment by MadhavJivrajani** [comment]
For the test `TestWithVerificationWithoutCaCertOrThumbprint`, we will also need to update the way the error is unwrapped here and account for the new `CertificateVerificationError`:

https://github.com/kubernetes/kubernetes/blob/ba0dbdbd11b063aca97121b6436141b37f5a1ec0/staging/src/k8s.io/legacy-cloud-providers/vsphere/vclib/connection_test.go#L185-L197

**Comment by MadhavJivrajani** [comment]
Created
- https://github.com/kubernetes/kubernetes/pull/114762
- https://github.com/kubernetes/kubernetes/pull/114763 (to be merged only after this is merged since it contains types introduced in go1.20).

**Comment by liggitt** [comment]
We need to keep track of all dependency and code changes associated with the update to go1.20, so we can pick/replay those changes on release branches and transition them to go1.20 as well once master is solid on it

**Comment by MadhavJivrajani** [comment]
/test pull-kubernetes-dependencies-go-canary
will fail due to the TODO in this PR
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
should pass after https://github.com/kubernetes/kubernetes/pull/114766 is merged, running to see if any more failures come up
/test pull-kubernetes-verify-go-canary
will fail still (see https://github.com/kubernetes/kubernetes/pull/114502#issuecomment-1364684294), running it again to make sure there aren't any new failures

Running go canaries again after bumping Go canary version to go1.20rc2, xref: https://github.com/kubernetes/test-infra/pull/28363

**Comment by liggitt** [comment]
/hold

needs rebase anyway, but we should wait to merge this at least until postsubmits run on the 1.19.5 merge

**Comment by MadhavJivrajani** [comment]
@liggitt yep
The verify job I think will only be green on 1.20 since go-tools makes a release only when go makes an official release (go1.20)

Ref https://github.com/kubernetes/kubernetes/pull/114502#issuecomment-1364684294

**Comment by liggitt** [comment]
we should be able to bump to HEAD of that tool and get a version that passes on go1.19 and go1.20rcX - opened https://github.com/kubernetes/kubernetes/pull/115129 to do this

**Comment by MadhavJivrajani** [comment]
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-verify-go-canary

Running canary jobs on account of the 1.20rc3 bump taking effect: https://github.com/kubernetes/test-infra/pull/28439

Verify job should still be failing (Jordan's PR should fix that)

**Comment by MadhavJivrajani** [comment]
/test pull-kubernetes-verify-go-canary

This should pass now that Jordan's PR is merged.

**Comment by liggitt** [comment]
I'm not actually even sure we need the pull-kubernetes-e2e-gce-canary run... the normal e2e run shows client and server built with go1.20rc3

```
I0118 07:02:36.982] Client Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.0-alpha.0.1050+146a7c8b95a725", GitCommit:"146a7c8b95a725042bb8efcc87191352b5942786", GitTreeState:"clean", BuildDate:"2023-01-18T05:36:35Z", GoVersion:"go1.20rc3", Compiler:"gc", Platform:"linux/amd64"}
I0118 07:02:36.982] Kustomize Version: v4.5.7
I0118 07:02:36.983] Server Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.0-alpha.0.1050+146a7c8b95a725", GitCommit:"146a7c8b95a725042bb8efcc87191352b5942786", GitTreeState:"clean", BuildDate:"2023-01-18T05:36:35Z", GoVersion:"go1.20rc3", Compiler:"gc", Platform:"linux/amd64"}
```

the unit / integration / verify / dependencies go-canary runs are still needed/useful, and those are green

**Review by liggitt — COMMENTED** [review]
one comment, then lgtm. not sure if we want to wait for the next go RC (since there's at least one revert on the go side that means 1.20rc3 won't match the final 1.20 content) or merge this earlier

**Comment by liggitt** [comment]
WIP to control the version of go used in presubmits via the .go-version file in https://github.com/kubernetes/kubernetes/pull/115377 turned up an issue with verify-typecheck using go1.20rc3 this PR missed (because we don't have a go-canary for verify-typecheck). Fix in https://github.com/kubernetes/kubernetes/pull/115382

**Comment by cpanato** [comment]
sorry for the delay @liggitt, i was sick and offline last week.

Do we want to merge this using the rc? or should we wait for the official release?

**Comment by liggitt** [comment]
let's get https://github.com/kubernetes/kubernetes/pull/115382 merged first

there's been a decent number of fixes since 1.20rc3 in https://github.com/golang/go/commits/release-branch.go1.20 ... I'd probably wait for the next rc to merge this

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/115377 is merged, can you rebase on latest master so the `.go-version` bump informs all presubmits?

**Comment by liggitt** [comment]
(once rebased on master, this lgtm)

**Comment by cpanato** [comment]
> #115377 is merged, can you rebase on latest master so the `.go-version` bump informs all presubmits?

rebased

**Pull request merged**